pkg/providers/vsphere/vmprovider_vm.go

```diff
@@ -1466,16 +1469,141 @@ func (vs *vSphereVMProvider) getVMYamlFromSnapshot(
+			// profile but it won't be part of the spec since the plumbing from VM class to VCFA won't exist (VCFA
+			// expects slots of a class to be reserved if this is set).
+			Reserved: nil,
+
+			// The VM will continue to have the same boot options from the snapshot state.
+			// If user wants to change that, they can specify a custom boot option (boot order).
+			BootOptions: nil,
+
+			// currentSnapshot is only used to trigger a revert, so this must be empty.
+			CurrentSnapshot: nil,
+
+			// Group name is overridden to the VM's current group.
+			// We can't know (or guess) what the group the VM was when the snapshot was taken.
+			GroupName: vmCtx.VM.Spec.GroupName,
+
+			// When Snapshot is taken with Policy A and then VM is updated to use Policy B,
+			// after a revert to Snapshot with Policy A is performed, VM continues to have Policy A.
+			StorageClass: vmCtx.VM.Spec.StorageClass,
+
+			Advanced: &vmopv1.VirtualMachineAdvancedSpec{
+				ChangeBlockTracking: moSnap.Config.ChangeTrackingEnabled,
+			},
+
+			Network: &vmopv1.VirtualMachineNetworkSpec{
+				HostName: vmCtx.MoVM.Guest.HostName,
+			},
+		}
+
+		// Inferred from the power state of the snapshot
+		// It doesn't matter what the power state of the VM is at the time of revert.
+		// The VM is always reverted in the same power state as the snapshot.
+		snapshot := vmlifecycle.FindSnapshotInTree(vmCtx.MoVM.Snapshot.RootSnapshotList, snap.Value)
+		if snapshot == nil {
+			// weird situation here
+			// the snapshot being processed doesn't exist in the snapshot tree
+			// this shouldn't be happening
+			err := fmt.Errorf("snapshot %s not found in VM Snapshot Tree", snap.Value)
+			vmCtx.Logger.Error(err, "")
+			return "", err
+		}
+		// convert from vimtypes.VirtualMachinePowerState to vmopv1.VirtualMachinePowerState
+		poweredOn := snapshot.State == vimtypes.VirtualMachinePowerStatePoweredOn
+		if poweredOn {
+			vm.Spec.PowerState = vmopv1.VirtualMachinePowerStateOn
+		} else {
+			vm.Spec.PowerState = vmopv1.VirtualMachinePowerStateOff
+		}
+
+		// Number of elements in the spec.network.interfaces list must be same as number of VM's NICs from ConfigInfo
+		// Details of each interface (which network does it connect to) is ignored
+		// Networking can be fixed after revert and it is not possible to map the vSphere network to Supervisor networks.
+		devices := object.VirtualDeviceList(moSnap.Config.Hardware.Device)
+		vNICs := devices.SelectByType((*vimtypes.VirtualEthernetCard)(nil))
+		if len(vNICs) == 0 {
+			// If no vNICs are defined, then disable network
+			vm.Spec.Network.Disabled = true
+		} else {
+			interfaces := make([]vmopv1.VirtualMachineNetworkInterfaceSpec, len(vNICs))
+			for i := range vNICs {
+				interfaces[i] = vmopv1.VirtualMachineNetworkInterfaceSpec{
+					Name: fmt.Sprintf("eth%d", i),
+				}
+			}
+			vm.Spec.Network.Interfaces = interfaces
+		}
+
+		// MinHardwareVersion is set to the Current Hardware Version of the VM
```

Comment:
Question, I assume moSnap stores hardware version of the snapshot when the snapshot is taken, other than the current hardware version of the VM?

Reply:
So it's the previous hardware version than to the current hardware version?